Support for Hex/Proficnc UAVCAN based Flow Sensor

bugobliterator opened the pull request.

rmackay9 commented:
I added the DevCallTopic but I don't think we actually got to it this week.

WickedShell commented on libraries/AP_RangeFinder/RangeFinder.cpp:
```diff
@@ -325,7 +326,18 @@ void RangeFinder::init(enum Rotation orientation_default)
+    if (num_instances < RANGEFINDER_MAX_INSTANCES) {
+        if (drivers[num_instances] == nullptr && params[num_instances].type == RangeFinder_TYPE_UAVCAN) {
+            //keep looking until found
+            drivers[num_instances] = AP_RangeFinder_UAVCAN::detect(state[num_instances], params[num_instances]);
```
If you failed to allocate it the first time, how is this likely to succeed?

WickedShell commented on libraries/AP_RangeFinder/RangeFinder.cpp:
```diff
@@ -325,7 +326,18 @@ void RangeFinder::init(enum Rotation orientation_default)
+            //keep looking until found
+            drivers[num_instances] = AP_RangeFinder_UAVCAN::detect(state[num_instances], params[num_instances]);
+            if (drivers[num_instances] != nullptr) {
+                num_instances++;
```
This doesn't honor the contract of `num_instances` being the highest detected instance correctly, and can corrupt other sensors.

WickedShell replied:
While that sounds reasonable, it actually isn't the API we have in a lot of libraries. `num_instances` is the number of the highest detected backend. If you don't setup rangefinder 1, but do setup 2 and detect a driver on 2, then `num_instances` will be 2. https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_RangeFinder/RangeFinder.cpp#L312

The problem here is if someone leaves an empty slot, then things will go wrong.

WickedShell commented on libraries/AP_UAVCAN/dsdl/com/hex/equipment/flow/20200.Measurement.uavcan:
```diff
@@ -0,0 +1,4 @@
+float32 integration_interval
```
Units? (on all of these)

rmackay9 commented on libraries/AP_OpticalFlow/OpticalFlow.cpp:
```diff
@@ -154,7 +154,10 @@ void OpticalFlow::update(void)
         return;
     }
-
+    if (backend == nullptr && (OpticalFlowType)_type.get() == OpticalFlowType::UAVCAN) {
```
We discussed this on the dev call and @tridge suggested that if the _TYPE = UAVCAN then we should always create the driver but it can be unhealthy which should then be caught in the pre-arm checks.

WickedShell commented on libraries/AP_OpticalFlow/AP_OpticalFlow_UAVCAN.cpp:
```diff
@@ -0,0 +1,110 @@
+    }
+    if (_driver == nullptr) {
+        _driver = new AP_OpticalFlow_UAVCAN(flow);
+        if (_driver == nullptr) {
```
If we failed to allocate the memory the first time, why are we likely to the second time? Given our use case this seems like an excessive usage of code.

rmackay9 commented:
As a side note, it would be good to have the PR description filled out with info including how much testing has been done.

rmackay9 commented:
I've test flown this and it worked quite well ([video](https://www.youtube.com/watch?v=MKJB_7cA_0s))

I haven't tested the range of the lidar but I can see it works.

WickedShell commented on libraries/AP_RangeFinder/RangeFinder.h:
```diff
@@ -35,7 +35,7 @@ class AP_RangeFinder_Backend;
 {
     friend class AP_RangeFinder_Backend;
-
+    friend class AP_RangeFinder_UAVCAN;
```
Why do you need a friend class? Just can't see what you need/that you aren't getting via the backend?

bugobliterator replied:
I need it to get list of drivers, I scan through them and initialise in the backend itself.

bugobliterator replied:
I will place a comment for reference.

WickedShell replied:
But why not allocated them in the loop in RangeFinder.cpp like normal ones, then you can simply leave them with an unassigned flag? This resolves a couple of things for you:

- `num_instances` has been dropped from your patches, which means it won't be reporting correctly for UAVCAN that I can see (which is a real bug)
- Your fix of adding a backend type flag works, but if you allocate the drivers here, you don't need to introduce the extra member everywhere, and only pay the cost of the bool flag on the UAVCAN instances
- It would allow you to remove the friend class, and use the param instance you are already passing to the backend

bugobliterator replied:
Yes, you are right. Except I will still need to be friends with frontend to get the list of drivers unless I maintain a list in the backend itself.

WickedShell replied:
Yeah that's true. Each driver would have to register itself onto a static list, which is going to be messy as well... (Would get you out of friends again though :D)

Either way `num_instances` needs help :)

bugobliterator replied:
Yes, num_instances was a real bug, thanks for pointing it out. Fixed in https://github.com/ArduPilot/ardupilot/pull/11017/commits/14c191bd4a8eea0007274009c8e92386515258fd

WickedShell commented on libraries/AP_RangeFinder/AP_RangeFinder_UAVCAN.cpp:
```diff
@@ -0,0 +1,136 @@
+
+void AP_RangeFinder_UAVCAN::update()
+{
+    WITH_SEMAPHORE(_sem);
```
Nevermind, I found it, forgot that all backends get one...

bugobliterator commented on libraries/AP_RangeFinder/AP_RangeFinder_UAVCAN.cpp:
```diff
@@ -0,0 +1,136 @@
+            driver = (AP_RangeFinder_UAVCAN*)AP::rangefinder().drivers[i];
+        }
+        if (driver != nullptr &&
+            driver->ap_uavcan == ap_uavcan && 
```
All three params won't match during the change, i.e. ap_uavcan and node_id. So, we will be fine. The driver won't ever be used, unless rebooted and reinitialised as UAVCAN backend.

bugobliterator replied:
Should be fixed in this patch: https://github.com/ArduPilot/ardupilot/pull/11017/commits/0e682b6b7302870437ddd132ede7a383c8b98c88

bugobliterator commented on libraries/AP_RangeFinder/AP_RangeFinder_UAVCAN.cpp:
```diff
@@ -0,0 +1,136 @@
+        update_status();
+        _new_data = false;
+    } else {
+        set_status(_status);
```
Fixed in this patch: https://github.com/ArduPilot/ardupilot/pull/11017/commits/c275a026928e16feb07908ac1536da48dc309520

WickedShell replied:
This is still stroking the `range_valid_count` to rapidly. You did actually want the `new_data` flag you had.

Basically this should loosely look like:
```
if new_data
  copy data
  update_status()
else if millis() - last_reading > 200
  set_status(NoData)
```

With nothing being done here if one of those 2 conditions aren't true.

bugobliterator replied:
Fixed with this: https://github.com/ArduPilot/ardupilot/pull/11017/commits/1e5fc1ef802c2c6a5fdf79440077c18fddf5a0da

bugobliterator replied:
Needed to add status set for other states being sent by sensor.

bugobliterator commented on libraries/AP_RangeFinder/AP_RangeFinder_UAVCAN.cpp:
```diff
@@ -0,0 +1,136 @@
+    if (driver == nullptr) {
+        return;
+    }
+    {
```
Fixed in new set of patches

bugobliterator commented on libraries/AP_RangeFinder/AP_RangeFinder_UAVCAN.h:
```diff
@@ -0,0 +1,45 @@
+    uint16_t _distance_cm;
+    uint16_t _last_reading_ms;
+    bool _new_data;
+    AP_UAVCAN* _ap_uavcan;
```
Yes, fixed now

bugobliterator commented on libraries/AP_RangeFinder/AP_RangeFinder_UAVCAN.h:
```diff
@@ -0,0 +1,45 @@
+    uint16_t _last_reading_ms;
+    bool _new_data;
+    AP_UAVCAN* _ap_uavcan;
+    uint8_t _node_id;
```
Yes, Fixed Now

bugobliterator commented on libraries/AP_RangeFinder/AP_RangeFinder_UAVCAN.h:
```diff
@@ -0,0 +1,45 @@
+    AP_UAVCAN* ap_uavcan;
+    uint8_t node_id;
+
+    static HAL_Semaphore _sem_registry;
```
Yes, Fixed.

bugobliterator commented on libraries/AP_RangeFinder/AP_RangeFinder_UAVCAN.h:
```diff
@@ -0,0 +1,45 @@
+        return MAV_DISTANCE_SENSOR_LASER;
+    }
+private:
+    static bool take_registry();
```
Yes, Fixed

bugobliterator commented on libraries/AP_RangeFinder/AP_RangeFinder_UAVCAN.h:
```diff
@@ -0,0 +1,45 @@
+    }
+private:
+    static bool take_registry();
+    static void give_registry();
```
Yes, Fixed

bugobliterator commented on libraries/AP_RangeFinder/AP_RangeFinder_UAVCAN.cpp:
```diff
@@ -0,0 +1,136 @@
+ */
+AP_RangeFinder_UAVCAN::AP_RangeFinder_UAVCAN(RangeFinder::RangeFinder_State &_state, AP_RangeFinder_Params &_params) :
+    AP_RangeFinder_Backend(_state, _params),
+    _new_data(false)
```
Yes, Fixed

bugobliterator commented on libraries/AP_RangeFinder/AP_RangeFinder_UAVCAN.h:
```diff
@@ -0,0 +1,45 @@
+
+protected:
+    virtual MAV_DISTANCE_SENSOR _get_mav_distance_sensor_type() const override {
+        return MAV_DISTANCE_SENSOR_LASER;
```
Fixed with this commit: https://github.com/ArduPilot/ardupilot/pull/11017/commits/d3599c74bcb8e2f023f912a56fe9b7418a24e2bc

WickedShell commented on libraries/AP_UAVCAN/dsdl/com/hex/equipment/flow/20200.Measurement.uavcan:
```diff
@@ -0,0 +1,4 @@
+float32 integration_interval    # radians
+float32[2] rate_gyro_integral   # radians
+float32[2] flow_integral        # seconds
+uint8 quality
```
How is this interpreted?

bugobliterator replied:
There's a mistake there I will fix it. Needs to be seconds, radians, radians. I will also add more details.

bugobliterator replied:
Fixed in https://github.com/ArduPilot/ardupilot/pull/11017/commits/37490b4e60f55834e319499a24f546eaf954f08e

WickedShell replied:
What's better 0 or 255?

bugobliterator replied:
Fixed, thanks

WickedShell commented on libraries/AP_OpticalFlow/AP_OpticalFlow_HereFlow.cpp:
```diff
@@ -0,0 +1,102 @@
+void AP_OpticalFlow_HereFlow::handle_measurement(AP_UAVCAN* ap_uavcan, uint8_t node_id, const MeasurementCb &cb)
+{
+    if (_driver == nullptr) {
+        _ap_uavcan = ap_uavcan;
```
This is the only usage of this pointer, and it's a write only. Why do you need to store the pointer at all?

bugobliterator replied:
Fixed in https://github.com/ArduPilot/ardupilot/pull/11017/commits/6970978f1237353130863dd087089047a0a5c93a

WickedShell commented on libraries/AP_OpticalFlow/AP_OpticalFlow_HereFlow.cpp:
```diff
@@ -0,0 +1,102 @@
+{
+    if (_driver == nullptr) {
+        _ap_uavcan = ap_uavcan;
+        _node_id = node_id;
```
Similarly this looks to be write only.

bugobliterator replied:
Fixed in https://github.com/ArduPilot/ardupilot/pull/11017/commits/6970978f1237353130863dd087089047a0a5c93a

rmackay9 commented:
I'm keen to get this into master so I've added the DevCallTopic so that it will go through a review during next week's dev call.  I think it would be nice to get it cleaned up a bit before then (ie. the items in my review).  Thanks!

WickedShell commented:
@rmackay9 I'm almost guaranteed to not be at the call due to the AUVSI show unfortunately.

WickedShell commented on libraries/AP_RangeFinder/AP_RangeFinder_UAVCAN.cpp:
```diff
@@ -0,0 +1,167 @@
+    WITH_SEMAPHORE(_sem);
+    if ((AP_HAL::millis() - _last_reading_ms) > 200) {
+        set_status(RangeFinder::RangeFinder_NoData);
+        state.distance_cm = 0;
```
Why bother with this? We should be using the status to guard it, making this a throw away value. (And if isn't guarded this is also a problematic number)

WickedShell commented on libraries/AP_RangeFinder/AP_RangeFinder_UAVCAN.cpp:
```diff
@@ -0,0 +1,167 @@
+        state.distance_cm = _distance_cm;
+        state.last_reading_ms = _last_reading_ms;
+        update_status();
+        new_data = true;
```
This should be clearing to false.

WickedShell commented on libraries/AP_RangeFinder/AP_RangeFinder_UAVCAN.cpp:
```diff
@@ -0,0 +1,167 @@
+            driver->_distance_cm = cb.msg->range*100.0f;
+            driver->_last_reading_ms = AP_HAL::millis();
+            driver->_status = RangeFinder::RangeFinder_Good;
+            driver->new_data = false;
```
this should be setting true

WickedShell commented:
Based on the 2 errors I just found I suspect this hasn't been tested against real hardware after all the other changes, as it looks like this would have never passed a valid piece of range data in.

tridge commented:
can you squash to min number of patches?

bugobliterator commented:
@tridge I have squashed the commits.

rmackay9 commented:
@bugobliterator, we should remove the EKF2 change..

bugobliterator commented:
> @bugobliterator, we should remove the EKF2 change..

Done.

rmackay9 commented:
@bugobliterator, I was wondering if you've re-tested this on a real vehicle?  I'm just referring to the two issues that @WickedShell found that indicate it may not have been tested.

bugobliterator commented:
> @bugobliterator, I was wondering if you've re-tested this on a real vehicle? I'm just referring to the two issues that @WickedShell found that indicate it may not have been tested.

@rmackay9 it is now tested working. I forgot to test the RangeFinder, so my bad on that. RangeFinder is now working with latest changes as well.

rbachtell commented:
I can verify the functionality of the code, I have two vehicles currently flying with OF and RF working as designed.

rmackay9 commented:
@bugobliterator,

I'm seeing Bad LiDAR Health a lot in the MP during testing.  It's intermittent but the message appear very frequently (for me at least).

I think it's not a good idea to be setting the _status both within the handle_measurement() function and then again within the update() function.  I think it makes it more difficult to understand which functions are responsible and leads to bugs.  Also I think there is a real issue - for example, I think handle_measurement could set the _status to Good because the sensor says it got a good reading, but this reading go be further away than the RNGFND1_MAX_CM value.  In these cases the _status should really be set to RangeFinder_OutOfRangeHigh.

By the way, I don't think you've gone through and added any additional comments.  It's not a blocking requirement but I think this driver has perhaps less comments than any other driver we have so it's a bit of the norm.

rbachtell commented:
I did have bad lidar health on one of my vehicles but in a conversation with Philip he looked at my log and said that at this point the lidar was designed for low altitude well lit indoor use and that outdoors could be problematic. I took this to be be more of a hardware issue than firmware so disabled the rangefinder and moved on.  It does make sense to have an out of range high warning though instead of bad health. Might be worth getting Philip involved, if he isn’t already, just to be sure that we’re not fighting a hardware limitation. I’ll watch the progress and when changes come up I’ll continue testing. 

Randy Bachtell
Sent from my iPad

> On May 9, 2019, at 6:51 PM, Randy Mackay <notifications@github.com> wrote:
> 
> @bugobliterator,
> 
> I'm seeing Bad LiDAR Health a lot in the MP during testing. It's intermittent but the message appear very frequently (for me at least).
> 
> I think it's not a good idea to be setting the _status both within the handle_measurement() function and then again within the update() function. I think it makes it more difficult to understand which functions are responsible and leads to bugs. Also I think there is a real issue - for example, I think handle_measurement could set the _status to Good because the sensor says it got a good reading, but this reading go be further away than the RNGFND1_MAX_CM value. In these cases the _status should really be set to RangeFinder_OutOfRangeHigh.
> 
> By the way, I don't think you've gone through and added any additional comments. It's not a blocking requirement but I think this driver has perhaps less comments than any other driver we have so it's a bit of the norm.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.

bugobliterator commented:
> @bugobliterator,
> 
> I'm seeing Bad LiDAR Health a lot in the MP during testing. It's intermittent but the message appear very frequently (for me at least).
> 
I have updated the code to handle undefined as no data received and let top level method to have timeout based estimate on NoData. You might be getting lot of undefined data, that's why it behaved like that.

> I think it's not a good idea to be setting the _status both within the handle_measurement() function and then again within the update() function. I think it makes it more difficult to understand which functions are responsible and leads to bugs. Also I think there is a real issue - for example, I think handle_measurement could set the _status to Good because the sensor says it got a good reading, but this reading go be further away than the RNGFND1_MAX_CM value. In these cases the _status should really be set to RangeFinder_OutOfRangeHigh.

The RangeHigh and RangeLow limit are handled based on response from HW as well as Ardupilot's internal parameter setting. update_state handles the good range measurements to be bounded by MAX and MIN as well. In case user sets too high or too low values than HW can handle, the vehicle will be protected by bounds set by HW itself as well.

> By the way, I don't think you've gone through and added any additional comments. It's not a blocking requirement but I think this driver has perhaps less comments than any other driver we have so it's a bit of the norm.
I have updated the source codes with some comments now.

rmackay9 commented:
I think what may be happening is that two different threads (the main thread that runs the "update" and then the I/O thread that runs the **handle_measurement** are both updating the status.  I don't think we can do this because it will result in a race condition and there's no semaphores protecting us.

Maybe I'm wrong, perhaps another developer (like @OXINARF) will know better and say I'm wrong but for barring that, I think this needs to be handled like other range finder drivers (like the Benewake) where we average across readings while also handling feedback from the sensor saying that some readings are out of range.

rmackay9 commented:
I've tested this again and if I use @bugobliterator's branch (i.e. the branch from this PR) then the range finder seems to work OK.  If I rebase it off master though then I see "NoData" status from the range finder.  I've included log files below:

- [logfile from this PR's branch working OK](https://www.dropbox.com/s/59f12kfjbs3zq7s/sids-branch.bin?dl=0)
- [logfile from this PR's branch rebased on master showing NoData sometimes](https://www.dropbox.com/s/8xxxsr7sg0bye14/sids-rebased-on-master.bin?dl=0)

rbachtell commented:
I have compiled the code from latest master with the latest PR files copied over.  I am not a programmer and know just enough C++ to be dangerous.  That being said the OF and RF are working, the OF appears to be solid with good data.  The RF shows accurate data and I have set max range to 140 to be safely within limits, at this point I get consistent recurring bad lidar health messages with altitudes above max.  I'm unsure if this is code or hardware triggering the warning.  I have included a log that is primarily around the max altitude set point and one that has some time well above. 
00000156.BIN
00000157.BIN

rmackay9 commented:
The flow sensor itself is working well but the range finder doesn't seem to perform very well.  I suspect the rangefinder is going to be a source of user pain and support calls but it doesn't appear that much more than be done on the driver side to improve things so I'm going to merge this.
